pkg/templating/api.go

```diff
@@ -153,6 +167,72 @@ func (lo ParentLabelSetAdder) Patch(cr resource.ParentResource, list []resource.
 	return list, nil
 }
 
+// NewAPIOrderedDeleter returns a new *APIOrderedDeleter.
+func NewAPIOrderedDeleter(c client.Client) *APIOrderedDeleter {
+	return &APIOrderedDeleter{kube: c}
+}
+
+// APIOrderedDeleter deletes the child resources in an order that is determined
+// by their priority noted in the child resource annotation. The child resources
+// with higher priority will be deleted first and their deletion will block
+// the lower priority ones.
+type APIOrderedDeleter struct {
+	kube client.Client
+}
+
+// Delete executes an ordered deletion of child resources depending on their
+// deletion priority.
+func (d *APIOrderedDeleter) Delete(ctx context.Context, list []resource.ChildResource) ([]resource.ChildResource, error) {
+	hp := minInt
+	del := []resource.ChildResource{}
```

Comment:
The current implementation is fine, but I'm a little curious whether it would be possible to do this using `sort.Slice`. Something like:

```go
func priority(o metav1.Object) int {
	val, ok := res.GetAnnotations()[DeletionPriorityAnnotationKey]
	if !ok {
		return 0
	}
	p, err := strconv.Atoi(val)
	if err != nil {
		return 0
	}
	return p
}

sort.Slice(list, func(i, j resource.ChildResource) bool { return priority(i) < priority(j) }
```

Reply:
In one reconcile, we only operate on the highest level of priority that has some resources in the cluster, which can be extracted with one pass over the list. I thought about sorting in the beginning and then iterate over the sorted array but bookkeeping seemed similar to the existing one. But I'll give it a shot to see if the code reads better using sort.

Reply:
It ended up a little more complex with the bookkeeping and there wasn't a good way to incorporate the return of error when the annotation value is not an integer, so I decided to keep it as is.